rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/DynamicCapabilities.java

```diff
@@ -0,0 +1,237 @@
+        this.clientCapabilities = clientCapabilities;
+        this.serverCapabilities = serverCapabilities;
+
+        for (var cap : supportedCapabilities) {
+            if (cap.setStaticCapability(clientCapabilities, serverCapabilities)) {
+                staticCapabilities.add(cap);
+            }
+        }
+    }
+
+    public CompletableFuture<Void> registerCapabilities(ILanguageContributions contribs) {
+        logger.debug("Registering some capabilities");
+
+        // Compute registrations purely based on contributions
+        // This requires waiting for an evaluator to load, which might take long, and should not block our logbook
+        return CompletableFutureUtils.reduce(supportedCapabilities.stream().map(c -> registration(c, contribs)))
+            .thenAccept(capabilities -> {
+                // Since we have some bookkeeping to do, we will now block our logbook for a moment
+                synchronized (currentRegistrations) {
+                    List<Registration> registrations = new LinkedList<>();
+                    List<Unregistration> unregistrations = new LinkedList<>();
+                    for (var entry : capabilities) {
+                        var cap = entry.getLeft();
+                        var registration = entry.getRight().orElse(null);
+                        if (registration == null || staticCapabilities.contains(cap)) {
+                            continue;
+                        }
+
+                        // Check if we already have this registration
+                        var existingRegistration = currentRegistrations.get(registration.getMethod());
+                        if (existingRegistration != null) {
+                            if (Objects.deepEquals(registration.getRegisterOptions(), existingRegistration.getRegisterOptions())) {
+                                logger.trace("No option changes for {}", registration.getMethod());
+                                continue;
+                            }
+
+                            // Otherwise, compute merged options
+                            var unregistration = computeOptionChanges(registration, existingRegistration.getRegisterOptions(), cap);
+
+                            logger.trace("Adding dynamic capability {} to task list", registration.getMethod());
+                            if (unregistration != null) {
+                                unregistrations.add(unregistration);
+                            }
+                        }
+                        registrations.add(registration);
+                    }
+
+                    try {
+                        doRegistrations(registrations, unregistrations);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    } catch (ExecutionException e) {
+                        logger.error("Doing registrations failed!", e);
```

Comment:
It might be helpful to log which registration failed.

Reply:
execution exception is normally a completable future phase that threw an exception.